lib/Frontend/CompilerInvocation.cpp

```diff
@@ -1701,7 +1701,25 @@ static void ParseLangArgs(LangOptions &Opts, ArgList &Args, InputKind IK,
   Opts.PascalStrings = Args.hasArg(OPT_fpascal_strings);
   Opts.VtorDispMode = getLastArgIntValue(Args, OPT_vtordisp_mode_EQ, 1, Diags);
   Opts.Borland = Args.hasArg(OPT_fborland_extensions);
-  Opts.CheckedC = Args.hasArg(OPT_fcheckedc_extension);
+  if (Args.hasArg(OPT_fcheckedc_extension)) {
+    const char *disallowed = nullptr;
+    if (Opts.CUDA)
+      disallowed = "CUDA";
+    else if (Opts.OpenCL)
+      disallowed = "OpenCL";
+    else if (Opts.CPlusPlus)
+      disallowed = "C++";
+    else if (Opts.ObjC1 || Opts.ObjC2)
+      disallowed = "Objective C";
+    if (disallowed) {
+      Diags.Report(diag::err_drv_argument_not_allowed_with) <<
+        "-fcheckedc-extension" << disallowed;
+    } else if (!(Opts.C99 || Opts.C11)) {
+      Diags.Report(diag::err_drv_argument_only_allowed_with) <<
+        "-fcheckedc-extension" << "C";
```

Comment:
This is a bug.   I was trying to protect against someone adding a new major language extension by specifically calling out that only C was supported.  However, I misread the code - there is no way to say we "only support this  compiler flag with plain C" in the current system.   

After studying the code more carefully, I realized that clang assumes that ANSI C (C89) is the default and then builds upon that with various flags for C++ and newer versions of the C and C++ standards. We do want to support  to C89.  The fix is to remove the specific checks for C99 and C11.

For the way that clang represents language options, see clang\include\clang\Basic\LangOptions.def.    This is included by LangOptions.h, in the same directory.

The way clang internally represents language options is different from the default language version that is chosen if none is specified.   The driver currently sets the default language version for C to C11 with gnu extensions.